src/github/mod.rs

```diff
@@ -235,6 +257,48 @@ impl DefaultGithubClient {
         // kick off the retry loop
         retry(Self::backoff_config(), operation).await
     }
+
+    async fn rate_limit_guard(&self) {
+        let mut state = self.rate_limit.lock().await;
+
+        // define a safety threshold
+        let threshold = 10;
+        if state.remaining <= threshold {
+            let now = Instant::now();
+            if now < state.reset_at {
+                let wait = state.reset_at - now;
+                tracing::info!(
+                    "Approaching rate limit ({} left). Sleeping {:?} until reset...",
+                    state.remaining,
+                    wait
+                );
+                tokio::time::sleep(wait).await;
+            }
+        }
+    }
+
+    fn update_rate_limit_from_headers(&self, headers: &HeaderMap) {
+        let mut state = futures::executor::block_on(self.rate_limit.lock());
+        if let (Some(rem), Some(reset)) = (
```

Comment:
what if it fails to parse or name of headers changes? It should not fail silently